test/cmp_protect_test.c

```diff
@@ -0,0 +1,535 @@
+    int res = TEST_ptr(protection)
+                  && TEST_true(ASN1_STRING_cmp(protection,
+                                               fixture->msg->protection) == 0);
+
+    ASN1_BIT_STRING_free(protection);
+    return res;
+}
+
+/* This function works similar to parts of CMP_verify_signature in cmp_vfy.c,
+ * but without the need for a OSSL_CMP_CTX or a X509 certificate */
+static int verify_signature(OSSL_CMP_MSG *msg,
+                            ASN1_BIT_STRING *protection,
+                            EVP_PKEY *pkey, int digest_nid)
+{
+    CMP_PROTECTEDPART prot_part;
+    unsigned char *prot_part_der = NULL;
+    int l;
+    EVP_MD_CTX *ctx = NULL;
+    int res;
+
+    prot_part.header = OSSL_CMP_MSG_get0_header(msg);
+    prot_part.body = msg->body;
+    res =
+        TEST_int_ge(l = i2d_CMP_PROTECTEDPART(&prot_part, &prot_part_der), 0)
+            && TEST_ptr(ctx = EVP_MD_CTX_create())
+            && TEST_true(EVP_VerifyInit_ex
+                         (ctx, (EVP_MD *)EVP_get_digestbynid(digest_nid), NULL))
+            && TEST_true(EVP_VerifyUpdate(ctx, prot_part_der, l))
+            && TEST_int_eq(EVP_VerifyFinal(ctx, protection->data,
+                                           protection->length, pkey), 1);
+    /* cleanup */
+    EVP_MD_CTX_destroy(ctx);
+    OPENSSL_free(prot_part_der);
+    return res;
+}
+
+/* Calls OSSL_CMP_calc_protection and verifies signature*/
+static int execute_calc_protection_signature_test(CMP_PROTECT_TEST_FIXTURE *
+                                                  fixture)
+{
+    ASN1_BIT_STRING *protection =
+        ossl_cmp_calc_protection(fixture->msg, NULL, fixture->privkey);
+    int ret = (TEST_ptr(protection)
+                   && TEST_true(verify_signature(fixture->msg, protection,
+                                                 fixture->pubkey,
+                                                 fixture->cmp_ctx->digest)));
+
+    ASN1_BIT_STRING_free(protection);
+    return ret;
+}
+
+/* TODO TPa: find a way to set protection algorithm */
+static int test_cmp_calc_protection_no_key_no_secret(void)
+{
+    SETUP_TEST_FIXTURE(CMP_PROTECT_TEST_FIXTURE, set_up);
+    /* Do test case-specific set up; set expected return values and
+     * side effects */
+    if (!TEST_ptr(fixture->msg = load_pkimsg(ir_unprotected_f))
+         || !TEST_ptr(fixture->msg->header->protectionAlg = X509_ALGOR_new())) {
+        tear_down(fixture);
+        fixture = NULL;
+    }
+
+    EXECUTE_TEST(execute_calc_protection_fails_test, tear_down);
+    return result;
+}
+
+/* TODO TPa: find openssl-independent reference value */
+static int test_cmp_calc_protection_pkey(void)
+{
+    SETUP_TEST_FIXTURE(CMP_PROTECT_TEST_FIXTURE, set_up);
+    fixture->pubkey = loadedpubkey;
+    fixture->privkey = loadedprivkey;
+    if (!TEST_true(EVP_PKEY_up_ref(loadedpubkey))
+            || !TEST_true(EVP_PKEY_up_ref(loadedprivkey))
```

Comment:
Double free when we tear_down the fixture if an error occurs during up_ref (do we actually need to up_ref and then free in the tear_down...loadedpubkey/loadedprivkey will always live longer than the tests)

Reply:
You are right. I dropped the up_ref stuff for loadedpubkey and loadedprivkey.